python/KAPEL.py

```diff
@@ -53,6 +53,12 @@ def __init__(self, queryRange, namespace):
         # https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators
         # https://prometheus.io/docs/prometheus/latest/querying/operators/#many-to-one-and-one-to-many-vector-matches
         self.cputime = f'(max_over_time(kube_pod_completion_time{{namespace="{namespace}"}}[{queryRange}]) - max_over_time(kube_pod_start_time{{namespace="{namespace}"}}[{queryRange}])) * on (pod) group_left() max without (instance, node) (max_over_time(kube_pod_container_resource_requests{{resource="cpu", node != "", namespace="{namespace}"}}[{queryRange}]))'
+
+        # These are container-level memory and CPU usage metrics reported by kubelets.
+        self.memory = f'sum by (pod) (max_over_time(container_memory_working_set_bytes{{namespace="{namespace}"}}[{queryRange}])) / 1000'
```

Comment:
Unlike the container_cpu_usage_seconds_total which is a cumulative counter, the memory usage is an instantaneous gauge. For a pod with containers that have memory usage like this:
```
                                  minute 1              minute 2
container 1                    1 GB                          0 GB
container 2                    0 GB                          1 GB
```
the query would incorrectly report memory usage of 2 GB even though the real memory usage never exceeded 1 GB, because it is doing the sum of max_over_time of each. I tried to find a way to collapse together the range vector for all containers in a pod first, then take the max but couldn't figure it out. 

@mwestphall What would you think of using the combined resource memory request of all containers in a pod, like `sum by (pod) (max_over_time(kube_pod_container_resource_requests{resource="memory", node != "", namespace="harvester"}[1h]))` ?  This uses metrics from kube-state-metrics, which should be simpler and more reliable than the cgroup-based ones.

Besides, the requested memory is the usage in the sense of what a pod deprives other pods of using. The actual memory occupied by a process could be lower than what is requested, but the pod still "used" all the requested memory in the sense that it was reserved by the scheduler.

Reply:
@rptaylor Using the `kube_pod_container_resource_requests` metric in place of `container_cpu_usage_seconds_total` sounds good to me. 

Then with regards to multi-container support in general, we ran into an issue when we were first testing on tiger with the cputime query erroring out when run against multi-container pods. I believe this is because the `max without (instance, node)` grouping in the query doesn't always aggregate every container in a pod down to a single record. Would it make sense to a `max by (pod)` there instead?

Reply:
Do you mean using e.g. `sum by (pod) (max_over_time(kube_pod_container_resource_requests{resource="memory", node != "", namespace="harvester"}[1h]))` instead of container_memory_working_set_bytes ?

I think the container_cpu_usage_seconds_total query should be good now, after all the investigation we did here.

I opened a separate issue for that: https://github.com/rptaylor/kapel/issues/56

Reply:
I thought your commit "[report on pod total requested memory rather than max usage per container](https://github.com/rptaylor/kapel/pull/54/commits/a0e56e706bce6906b2958534f3d39431345e91b3)" was meant to address this but it looks like it only changes the chart name.

Reply:
Oops, good catch, that was definitely not the intended change with that commit. I've rebased that out and changed the `memory` query to `sum by (pod) (max_over_time(kube_pod_container_resource_requests{resource="memory", node != "", namespace="harvester"}[<queryRange>]))` as intended.